### PR TITLE
promql: Allow injecting fake tokens into the generated parser

### PR DESCRIPTION
**Pull request by slrtbtfs**
Yacc grammars do not support having multiple start symbols.

To work around that restriction, it is possible to inject fake tokens into the lexer stream,
as described here https://www.gnu.org/software/bison/manual/html_node/Multiple-start_002dsymbols.html .

This is part of the parser rewrite effort described in #6256.

Signed-off-by: Tobias Guggenmos <tguggenm@redhat.com>

### CODE REVIEW

**Comment by brian-brazil** [comment]
I'm confused, how is this even a problem?

**Comment by slrtbtfs** [comment]
Right now the parser already has several entry points:

* `ParseExpr()` accepts any valid PromQL expression.
* `ParseMetric()` accepts only a single metric. I guess its used for scraping metrics.
* `ParseMetricSelector()` accepts only a single metric and returns a set of label matchers.

So technically  the PromQL parser parses 3 different formal languages, that happen to share some syntax rules.

By using this start symbol trick, these can actually share the same grammar.

---

While the parser rewrite is in progress there will be even more entry points, e.g. one of metric selectors, so parts of the parsing can be done with a generated parser while the the old parser is used for the rest.

**Comment by slrtbtfs** [comment]
There already was something similar in the old generated parser: https://github.com/prometheus/prometheus/blob/0.10.0/rules/parser.y#L37

File: promql/lex.go [hunk 1]
```diff
@@ -193,6 +193,10 @@ const (
 	GROUP_RIGHT
 	BOOL
 	keywordsEnd
+
+	startSymbolsStart
+	// start symbols for the generated parser
```
Comment (brian-brazil) [inline promql/lex.go]: Capital letter and full stop.

Comment (slrtbtfs) [reply promql/lex.go]: Oh no, not again! :man_facepalming:

Comment (brian-brazil) [reply promql/lex.go]: Thought the bowl of petunias.

**Comment by brian-brazil** [comment]
Thanks!

**Pull request merged**